high-level.lisp

```diff
@@ -888,3 +888,35 @@ with upper left block with dimension P-by-Q. Returns the intermediate representa
     (flet ((calc-i-j (i j) (* (ref a (floor i mb) (floor j nb))
                               (ref b (mod i mb) (mod j nb)))))
       (reduce #'kron rest :initial-value (tabulate (* ma mb) (* na nb) #'calc-i-j)))))
+
+(defun num-to-bits (num)
+  (loop for i below (integer-length num)
+	collect (if (logbitp i num) 1 0)))
+
+(defun bin-decomp (num)
+  (let ((bits (num-to-bits num)))
+    (loop for i below (length bits)
+          for b in bits
+          when (> b 0)
+          collect (* b (expt 2 i)))))
+
+(defun exptm (m power)
```

Comment:
I think this function can be greatly simplified. The above two auxiliary functions can be deleted.

Here is a prototype of what `x^n` would look like for a simple number `x` and integer `n`:

```
(defun my-expt (x n)
  (labels ((recurse (y x n)
             (cond
               ;; x^(-n) = (1/x)^n
               ((minusp n) (recurse y (/ x) (- n)))
               ;; We are done recursing. Return answer.
               ((zerop n) y)
               ;; n is even, so square.
               ((evenp n) (recurse y (* x x) (ash n -1)))
               ;; n is odd, so multiply out and make it even.
               (t (recurse (* x y) x (1- n))))))
    (recurse 1 x n)))
```

You should be able to adapt this for matrices. Be mindful of the meaning of `1` and `/` in the context of matrices.

In fact, if we fix `x` to be a `single-float` and `n` to be a `fixnum`, as in the following code, we get a good disassembly.

```
(defun my-expt (x n)
  (declare (type single-float x)
           (type fixnum n)
           (optimize speed (safety 0) (debug 0) (space 0)))
  (labels ((recurse (y x n)
             (declare (type fixnum n)
                      (type single-float x y))
             (cond
               ;; x^(-n) = (1/x)^n
               ((minusp n) (recurse y (/ x) (- n)))
               ;; We are done recursing. Return answer.
               ((zerop n) y)
               ;; n is even, so square.
               ((evenp n) (recurse y (* x x) (ash n -1)))
               ;; n is odd, so multiply out and make it even.
               (t (recurse (* x y) x (1- n))))))
    (recurse 1.0f0 x n)))

;;;

CL-USER> (disassemble #'my-expt)

; disassembly for MY-EXPT
; Size: 112 bytes. Origin: #x23797C04
; 04:       F30F1015A4FFFFFF MOVSS XMM2, [RIP-92]             ; no-arg-parsing entry point
                                                              ; [#x23797BB0]
; 0C:       EB52             JMP L3
; 0E:       6690             NOP
; 10: L0:   4885FF           TEST RDI, RDI
; 13:       7C33             JL L2
; 15:       488BCF           MOV RCX, RDI
; 18:       4885C9           TEST RCX, RCX
; 1B:       7445             JEQ L4
; 1D:       488BCF           MOV RCX, RDI
; 20:       F6C102           TEST CL, 2
; 23:       7410             JEQ L1
; 25:       F30F59D0         MULSS XMM2, XMM0
; 29:       488BC7           MOV RAX, RDI
; 2C:       4883E802         SUB RAX, 2
; 30:       488BF8           MOV RDI, RAX
; 33:       EB2B             JMP L3
; 35: L1:   F30F59C0         MULSS XMM0, XMM0
; 39:       488BC7           MOV RAX, RDI
; 3C:       48D1F8           SAR RAX, 1
; 3F:       4883E0FE         AND RAX, -2
; 43:       488BF8           MOV RDI, RAX
; 46:       EB18             JMP L3
; 48: L2:   F30F100D60FFFFFF MOVSS XMM1, [RIP-160]            ; [#x23797BB0]
; 50:       F30F5EC8         DIVSS XMM1, XMM0
; 54:       0F28C1           MOVAPS XMM0, XMM1
; 57:       488BC7           MOV RAX, RDI
; 5A:       48F7D8           NEG RAX
; 5D:       488BF8           MOV RDI, RAX
; 60: L3:   EBAE             JMP L0
; 62: L4:   660F7ED2         MOVD EDX, XMM2
; 66:       48C1E220         SHL RDX, 32
; 6A:       4883CA19         OR RDX, 25
; 6E:       488BE5           MOV RSP, RBP
; 71:       F8               CLC
; 72:       5D               POP RBP
; 73:       C3               RET
```

Reply:
So translating that to matrices, it would be `(declare (type matrix x y))` and `(check-type x matrix)` etc?